File: src/ros2_bridge/mappers/geometry_msgs/polygon.rs
//! Typed mapper for `geometry_msgs/msg/Polygon`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn polygon_to_bus(
    msg: ros_env::geometry_msgs::msg::Polygon,
) -> crate::geometry_msgs::msg::v1::Polygon {
    crate::geometry_msgs::msg::v1::Polygon {
        points: msg
            .points
            .into_iter()
            .map(crate::ros2_bridge::mappers::geometry_msgs::point32::point32_to_bus)
            .collect(),
    }
}

pub(crate) fn polygon_to_ros(
    bus: crate::geometry_msgs::msg::v1::Polygon,
) -> ros_env::geometry_msgs::msg::Polygon {
    ros_env::geometry_msgs::msg::Polygon {
        points: bus
            .points
            .into_iter()
            .map(crate::ros2_bridge::mappers::geometry_msgs::point32::point32_to_ros)
            .collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsPolygonMapper;

impl TypedTopicMapper for GeometryMsgsPolygonMapper {
    type Ros = ros_env::geometry_msgs::msg::Polygon;
    type Bus = crate::geometry_msgs::msg::v1::Polygon;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(polygon_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(polygon_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/polygon.py
"""Generated mapper for `geometry_msgs/msg/Polygon`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.point32 import point32_to_bus, point32_to_ros

def polygon_to_bus(msg):
    from robot_bus.geometry_msgs.msg.v1 import Polygon as BusMsg

    bus = BusMsg()
    bus.points.extend([point32_to_bus(x) for x in msg.points])
    return bus


def polygon_to_ros(bus):
    from geometry_msgs.msg import Polygon as RosMsg

    out = RosMsg()
    out.points = [point32_to_ros(x) for x in bus.points]
    return out


class GeometryMsgsPolygonMapper:
    def ros_msg_type(self):
        from geometry_msgs.msg import Polygon as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return polygon_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.geometry_msgs.msg.v1 import Polygon as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return polygon_to_ros(bus)
